autocmd: introduce "++once" flag

justinmk opened the pull request:
Adds a new feature to :autocmd which sets the handler to be executed once, then removed.

Before:

    augroup FooGroup
      autocmd!
      autocmd FileType foo call Foo() | autocmd! FooGroup FileType foo
    augroup END

After:

    autocmd FileType foo ++once call Foo()

- commit 2 renames the `once` flag to `-once` and also adds support for `-nested`. This is to avoid yet another "flags" pattern not seen anywhere else in Vim, and instead be consistent with e.g. `:command` flags pattern.
- commit 3 renames again, to `++once` and `++nested`

justinmk committed:
autocmd: introduce "once" feature

Adds a new feature to :autocmd which sets the handler to be executed at
most one times.

Before:

    augroup FooGroup
      autocmd!
      autocmd FileType foo call Foo() | autocmd! FooGroup * <buffer>
    augroup END

After:

    autocmd FileType foo once call Foo()

justinmk committed:
autocmd: rename "once" and "nested" to hyphen-prefixed forms

Houl commented:
That's weird syntax.  I'd think  `:au <once> MyGroup BufRead * :DoSomething` would be more familiar, similar to `:h <nomodeline>`.  If `:-once` was a valid command, it would be hidden by your syntax.

justinmk commented:
If any of `MyGroup BufRead *` were valid commands then they also would be "hidden" as `:autocmd` struggles to decide which is an `augroup` and which is a valid command. It's all weird because `:autocmd` is weird. Hyphenated flags are slightly less weird and as mentioned consistent with `:command` itself.

So let's not get mired in hypotheticals, the goal is only to be slightly less inconsistent with Vim's predominant idioms.

brammool commented:
> If any of `MyGroup BufRead *` were valid commands then they also would
> be "hidden" as `:autocmd` struggles to decide which is an `augroup`
> and which is a valid command. It's all weird because `:autocmd` is
> weird. Hyphenated flags are slightly less weird and as mentioned
> consistent with `:command` itself.

Not many Ex commands use a dash for an option.  That's mainly because a
file name can start with a dash.

More commands use the ++option form.  Some the <option> form.
It's hard to decide what to use for new options to existing commands.
The ++option form appears to be the most common choice.  AFAIK it's
easier to type ++option than <option> on most keyboards.

-- 
hundred-and-one symptoms of being an internet addict:
51. You put a pillow case over your laptop so your lover doesn't see it while
    you are pretending to catch your breath.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

justinmk committed:
autocmd: rename: "++nested", "++once"

justinmk commented:
Renamed to `++once` and `++nested`.

Houl commented:
`:doautocmd` accepts the `<nomodeline>` argument (I'm repeating myself here).  For the sake of a bit of consistency, this should become `++nomodeline` then (?).  Alternatively: change `++once` to `<once>` and `++nested` to `<nested>`.  I think while `:command` is quite unrelated, `:autocmd` and `:doautocmd` are quite strongly related to each other.  Maybe this doesn't matter at all, just wanted to mention this.

justinmk commented:
> Alternatively: change ++once to `<once>` and ++nested to `<nested>`.

Angle-bracket variants are unpleasant to type as already mentioned.

ilAYAli commented:
> Renamed to `++once` and `++nested`.

hmm.. that looks like an obscure way to pre-increment the option to “twice”

Houl commented:
Angle-bracket variants are unpleasant to type as already mentioned.
Mentioned by someone who introduced this notation for `:doau`.  I'm fine with whatever choice, but this argument is silly.

justinmk committed:
autocmd: do not show empty section after ++once handlers expire

Problem: If autocmd pattern only contained `++once` handlers, and
         all of them completed, then there would be an empty group
         displayed by `:autocmd Foo`.
Solution: Delete the pattern if all of its commands were deleted.

justinmk commented:
Thanks, Bram!
